src/envoy/http/metadata_exchange/config.cc

```diff
@@ -34,22 +35,53 @@ using Common::Wasm::Null::Plugin::removeResponseHeader;
 using Common::Wasm::Null::Plugin::replaceRequestHeader;
 using Common::Wasm::Null::Plugin::replaceResponseHeader;
 
-void PluginContext::onCreate() {
-  // TODO(kuat): serialize metadata in the root context instead
-  // populate and encode node metadata
+void PluginRootContext::onConfigure(
+    std::unique_ptr<WasmData> ABSL_ATTRIBUTE_UNUSED configuration) {
   auto metadata =
       getMetadataValue(Common::Wasm::MetadataType::Node, NodeMetadataKey);
   if (metadata.kind_case() == google::protobuf::Value::kStructValue) {
     std::string metadata_bytes;
-    metadata.struct_value().SerializeToString(&metadata_bytes);
+    google::protobuf::io::StringOutputStream md(&metadata_bytes);
+    google::protobuf::io::CodedOutputStream mcs(&md);
+
+    mcs.SetSerializationDeterministic(true);
+    metadata.struct_value().SerializeToCodedStream(&mcs);
+
     metadata_value_ =
         Base64::encode(metadata_bytes.data(), metadata_bytes.size());
+
+    // magic "." to get the whole node.
+    auto node = getMetadataStruct(Common::Wasm::MetadataType::Node, ".");
+    for (auto& f : node.fields()) {
+      if (f.first == "id" &&
+          f.second.kind_case() == google::protobuf::Value::kStringValue) {
+        node_id_ = f.second.string_value();
+        break;
+      }
+    }
   }
 
   logDebug(
       absl::StrCat("metadata_value_ id:", id(), " value:", metadata_value_));
 }
 
+PluginRootContext* PluginContext::rootContext() {
+  return static_cast<PluginRootContext*>(this->root());
```

Comment:
s/static_cast/dynamic_cast (I did the same in the sd plugin based on review from yuchen :))

Reply:
Why? Is it creating a new copy ?

Reply:
My understanding was dynamic_cast is safer at downcast by type checking at runtime. But second thought static_cast actually works well here because `this->root()` should definitely return a `PluginRootContext` and could save a bit of runtime. So nvm about this.